// @sudoku/game.js
import { cursor } from './stores/cursor';
import { difficulty } from './stores/difficulty';
import { gamePaused } from './stores/game';
import { grid } from './stores/grid';
import { timer } from './stores/timer';
import { hints } from './stores/hints';
import { candidates } from './stores/candidates';
import { notes } from './stores/notes';

function resetSessionState() {
	cursor.reset();
	timer.reset();
	hints.reset();
	candidates.reset();
	notes.reset();
}

/**
 * 开始一局系统生成的新数独
 *
 * @param {('veryeasy' | 'easy' | 'medium' | 'hard')} diff - 难度
 */
export function startNew(diff) {
	difficulty.set(diff);
	grid.generate(diff);
	resetSessionState();

	location.hash = '';
}

/**
 * 使用自定义题面开始新游戏
 *
 * @param {string} sencode - 需要解码的题面字符串
 */
export function startCustom(sencode) {
	difficulty.setCustom();
	grid.decodeSencode(sencode);
	resetSessionState();
}

/**
 * 暂停游戏
 */
export function pauseGame() {
	timer.stop();
	gamePaused.set(true);
}

/**
 * 恢复游戏
 */
export function resumeGame() {
	timer.start();
	gamePaused.set(false);
}

export function undoMove() {
	// 历史记录操作交给接入领域层的 grid store 处理。
	grid.undo();
}

export function redoMove() {
	// 历史记录操作交给接入领域层的 grid store 处理。
	grid.redo();
}

export default {
	startNew,
	startCustom,
	undo: undoMove,
	redo: redoMove,
	pause: pauseGame,
	resume: resumeGame
};
